---
title: "Orebro RNA-seq analysis"
author: "Rodrigo Arcoverde Cerveira & Gustav Joas"
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
output: html_document
abstract: Ã–rebro study RNA-seq data analysis.
knit: (function(inputFile, encoding) {
          rmarkdown::render(inputFile,
                            encoding = encoding, 
                            output_file = paste0(
                              xfun::sans_ext(inputFile), '_', Sys.Date(), '.html'),
                                output_dir = "../results/lab_book/")})
editor_options: 
  chunk_output_type: inline
---

## Global knitr settings 

```{r, global.options, include=FALSE}
# Set up figures and chunks messages

knitr::opts_knit$set(
  echo = TRUE,
  root.dir = getwd(),
  fig.width = 6, fig.height = 5,
  fig.align = "center",
  out.width = 768,
  fig.pos = "H",
  warning = FALSE,
  message = FALSE
)
knitr::opts_chunk$set(
  warning = FALSE,
  message = FALSE,
  fig.width = 6, fig.height = 5,
  fig.align = "center",
  out.width = 768,
  fig.pos = "H"
)

result.dir <- paste0("results/", Sys.Date(), "/")
figures.dir <- paste0("results/", Sys.Date(), "/", "figures/")

# Create result.dir with date in if not existent
ifelse(isFALSE(dir.exists(paste0("../", result.dir))),
  dir.create(paste0("../", result.dir), recursive = TRUE),
  "Result directory for today exists already!"
)

ifelse(isFALSE(dir.exists(paste0("../", figures.dir))),
  dir.create(paste0("../", figures.dir), recursive = TRUE),
  "Result directory for today exists already!"
)

options(stringsAsFactors = FALSE)
```

## Loading libraries

```{r, results='hide'}
library(tidyr)
library(ggpubr)
library(ggsci)
library(ComplexHeatmap)
library(GeneOverlap)
library(data.table)
library(kableExtra)
library(forcats)
library(mgsub)
library(DESeq2)
library(pheatmap)
library(ggplot2)
library(ggrepel)
library(RColorBrewer)
library(limma)
library(edgeR)
library(enrichR)
library(gridExtra)
library(stringr)
library(dplyr)
library(ensembldb)
library(EnsDb.Hsapiens.v86)
library(factoextra)
library(rafalib)
library(tibble)
library(devtools)
library(clusterProfiler)
library(enrichplot)
library(ggprism)
library(msigdbr)
```



## Define directories

```{r, results='hide'}
# Define base directory
data_dir <- "../data/"

# Define gene counts directory
count_dir <- file.path(data_dir, "processed_data_nfcore/batch_1-2/salmon/")
## Gene counts file
count_file <- file.path(count_dir, "salmon.merged.gene_counts.tsv")

# Define metadata directory
sample_metadata_dir <- file.path(data_dir, "metadata/")

## Gene metadata file
gene_metadata_file <- file.path(sample_metadata_dir, "2023-11-222_mart_export_gene_ids_hsa.txt")
## Sample metadata file
sample_metadata_file <- file.path(sample_metadata_dir, "sample_metadata_1.csv")
## Patient metadata file
patient_metadata_file <- file.path(sample_metadata_dir, "patient_metadata.csv")
```

## Load and wrangle raw counts

```{r, results='hide'}
# Load raw counts
counts_raw <- data.table::fread(count_file,
  header = TRUE, sep = "\t"
) %>%
  as.data.frame() %>%
  select(-gene_id)

colnames(counts_raw) <- sub("_L001|_L002", "", colnames(counts_raw))
# Filter for only protein coding genes or mitochondrial/ribosomic genes
gene_id_metadata <- read.csv(gene_metadata_file) %>% filter(Gene.type %in% c("Mt_rRNA", "protein_coding", "rRNA"))
# Aggregate and sum up genes with same gene symbol (sum up isoform of genes)
counts_raw <- counts_raw[counts_raw$gene_name %in% gene_id_metadata$Gene.name, ]

counts_raw <- aggregate(counts_raw[, -1],
  list(gene_name = counts_raw[, 1]),
  FUN = sum
)

# Change row names to gene names
rownames(counts_raw) <- counts_raw$gene_name
counts_raw <- dplyr::select(counts_raw, -c(gene_name))
```

## Load metadata and match with raw counts

```{r, results='hide'}
# Load sample metadata
sampleTable <- read.table(sample_metadata_file,
  sep = ",", header = TRUE
) %>%
  dplyr::select(-c(age, sex))

# Load patient metadata
patient_metadata <- read.table(patient_metadata_file,
  sep = ",", header = TRUE
)

# Join sample and patient metadata
sampleTable <- sampleTable %>% left_join(patient_metadata, by = "subject_ID")

# Change visits to dose and time point...
sampleTable <- sampleTable %>%
  mutate(
    dose = as.factor(plyr::mapvalues(visit,
      from = c(1, 2, 4, 5, 10, 11, 12),
      to   = c("0_dose1", "24_dose1", "0_dose2", "24_dose2", "0_dose3", "24_dose3", "48_dose3")
    )),
    group = as.factor(group)
  ) %>%
  # ...(sex) 0 and 1 to male and female...
  mutate(
    sex = as.factor(plyr::mapvalues(sex,
      from = c(0, 1),
      to   = c("male", "female")
    )),
    group = as.factor(sex)
  ) %>%
  # ...(prior covid) 0 and 1 to neg and pos
  mutate(
    prior_covid19 = as.factor(plyr::mapvalues(prior_covid19,
      from = c(0, 1),
      to   = c("neg", "pos")
    )),
    group = as.factor(prior_covid19)
  )

# Remove 24dose_2 from subject 21 (missing 0dose_2)
sampleTable <- subset(sampleTable, sample_ID != "P22761_1037_S37")

# Remove one sample from subject 7 (sample replicate)
sampleTable <- subset(sampleTable, sample_ID != "P26208_1060_S62")

# Remove corresponding samples from count table
counts_raw <- subset(counts_raw, select = -c(P26208_1060_S62, P22761_1037_S37))

# Match count data with sample table
colnames(counts_raw) <- sampleTable$sample_ID
counts_raw <- counts_raw[, pmatch(sampleTable$sample_ID, colnames(counts_raw))]
all(rownames(sampleTable$sample_ID) == colnames(counts_raw))
```

## Basic Quality Control

### Inspect raw data

Plotting raw counts for first visual inspection. The boxplot shows median values of zero across all samples. In the histogram there is a huge peak of zero counts. Both indicate that the data set would benefit from a low count filtering.

```{r}
# Visualize distribution of raw counts with boxplot and density plot
boxplot(log2(as.matrix(counts_raw) + 1),
  ylab = expression("Log"[2] ~ "Read counts"),
  las = 2,
  main = "Raw data"
)
hist(log2(as.matrix(counts_raw) + 1),
  ylab = "",
  las = 2,
  main = "Raw data"
)
```

```{r, echo=FALSE, results='hide'}
# Saving plots
{
  pdf(paste0("../", figures.dir, "raw_counts_QC.pdf"),
    width = 10, height = 8, compress = TRUE
  )
  rafalib::mypar(1, 2, mar = c(10, 3, 3, 2))
  boxplot(
    outline = FALSE, log2(as.matrix(counts_raw) + 1),
    ylab = expression("Log"[2] ~ "Read counts"),
    las = 2,
    main = "Raw data"
  )
  hist(log2(as.matrix(counts_raw) + 1),
    ylab = "",
    las = 2,
    main = "Raw data"
  )
  par(mfrow = c(1, 1))
  garbage <- dev.off()
}
```

### Filtering

Plot number of genes detected across samples. All samples are more or less close to average so we don't need to discard any samples.

```{r}
barplot(colSums(counts_raw > 3),
  ylab = "Number of detected genes",
  las = 2
) +
  abline(h = median(colSums(counts_raw > 3)))
```

```{r, echo=FALSE, results='hide'}
# Saving plots
{
  pdf(paste0("../", figures.dir, "number_detected_genes.pdf"),
    width = 10, height = 8, compress = TRUE
  )
  par(mar = c(10, 4, 3, 4))
  barplot(colSums(counts_raw > 3),
    ylab = "Number of detected genes",
    las = 2
  ) +
    abline(h = median(colSums(counts_raw > 3)))
  dev.off()
}
```

Removing reads with the log2 of the counts per million (cpm) lower than 1 (= 45896 genes). Plot genes detection rate across samples, comparing raw and filtered counts.

```{r}
meanLog2CPM <- rowMeans(log2(cpm(counts_raw) + 1))
counts_filtered <- counts_raw[meanLog2CPM > 1, ]

hist(rowSums(counts_raw > 3), main = title("Raw Counts"))
hist(rowSums(counts_filtered > 3), main = title("Filtered Counts"))
```

```{r, echo=FALSE, results='hide'}
# Saving plots
{
  pdf(paste0("../", figures.dir, "detection_rate_raw_filtered.pdf"),
    width = 10, height = 8, compress = TRUE
  )
  par(mar = c(10, 4, 3, 4))
  hist(rowSums(counts_raw > 3), main = title("Raw Counts"))
  hist(rowSums(counts_filtered > 3), main = title("Filtered Counts"))
  par(mfrow = c(1, 1))
  dev.off()
}
```

Plot distribution of the filtered counts

```{r}
boxplot(log2(as.matrix(counts_filtered) + 1),
  ylab = expression("Log"[2] ~ "Read counts"),
  las = 2,
  main = "Filtered data"
)
hist(log2(as.matrix(counts_filtered) + 1),
  ylab = "",
  las = 2,
  main = "Filtered data"
)
```

```{r, echo=FALSE, results='hide'}
{
  pdf(paste0("../", figures.dir, "filtered_counts_distribution.pdf"),
    width = 10, height = 8, compress = TRUE
  )
  rafalib::mypar(1, 2, mar = c(10, 3, 3, 2))
  boxplot(log2(as.matrix(counts_filtered) + 1),
    ylab = expression("Log"[2] ~ "Read counts"),
    las = 2,
    main = "Filtered data"
  )
  hist(log2(as.matrix(counts_filtered) + 1),
    ylab = "",
    las = 2,
    main = "Filtered data"
  )
  par(mfrow = c(1, 1))
  dev.off()
}
```

### DESeq object creation

Generating three separate DESeq objects with different design parameters to allow for more comparisons. Controlling for age and sex.

```{r, results='hide'}
# Remove vaccine/sars-cov-2 related genes
counts_filtered <- counts_filtered[!grepl("vaccine|spike|sars_cov", rownames(counts_filtered)), ]

# Prepare for DESeq by creating new variables and factorizing
sampleTable <- sampleTable %>%
  mutate(
    conditions = paste0(sampleTable$group, "_", sampleTable$dose),
    time = gsub(".{6}$", "", conditions),
    conditions = factor(conditions),
    time = factor(time),
    sex = factor(sex),
    dose = factor(dose, levels = c("0_dose1", "24_dose1", "0_dose2", "24_dose2", "0_dose3", "24_dose3", "48_dose3")),
    group = factor(group),
    batch = factor(batch)
  )

# Create DESeq objects with different designs; comparing against conditions, time and dose
dds1 <- DESeqDataSetFromMatrix(
  countData = round(counts_filtered),
  colData = as.data.frame(sampleTable),
  design = ~ sex + age + batch + conditions
)

dds2 <- DESeqDataSetFromMatrix(
  countData = round(counts_filtered),
  colData = as.data.frame(sampleTable),
  design = ~ sex + age + batch + time
)

dds3 <- DESeqDataSetFromMatrix(
  countData = round(counts_filtered),
  colData = as.data.frame(sampleTable),
  design = ~ sex + age + batch + dose
)

dds4 <- DESeqDataSetFromMatrix(
  countData = round(counts_filtered),
  colData = as.data.frame(sampleTable),
  design = ~ sex + age + batch + dose + group + dose:group
)
```

### Normalization and Quality Control

Normalize with variance stabilizing transformation Plot distribution of data after normalization

```{r}
# Normalize with variance stabilizing transformation for later PCA and heatmap
counts_vst_normalized <- vst(dds1, blind = TRUE)

vst_matrix <- assay(counts_vst_normalized)
hist(vst_matrix)
boxplot(vst_matrix,
  ylab = expression("Log"[2] ~ "Read counts"),
  las = 2,
  main = "VST"
)
```

```{r, echo=FALSE, results='hide'}
# Saving plots

{
  pdf(paste0("../", figures.dir, "VST.pdf"),
    width = 10, height = 8, compress = TRUE
  )
  rafalib::mypar(1, 2, mar = c(10, 3, 3, 2))
  hist(vst_matrix)
  boxplot(vst_matrix,
    ylab = expression("Log"[2] ~ "Read counts"),
    las = 2,
    main = "VST"
  )
  par(mfrow = c(1, 1))
  dev.off()
}
```

### Heatmap

```{r}
# Sample heatmap
sampleDist <- cor(vst_matrix, method = "spearman")

Metadata <- data.frame(sampleTable$group, sampleTable$dose)
names(Metadata) <- c("Group", "Dose")
rownames(Metadata) <- sampleTable$sample_ID

# Plot heatmap
colors <- colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(255)

Heatmap <- pheatmap(sampleDist,
  color = colors,
  clustering_distance_rows = as.dist(1 - sampleDist),
  clustering_distance_cols = as.dist(1 - sampleDist),
  show_rownames = FALSE,
  show_colnames = FALSE,
  clustering_method = "ward.D2",
  annotation_col = Metadata
)

print(Heatmap)
```

```{r, echo=FALSE}
# Saving heatmap
ggsave(
  filename = paste0("../", figures.dir, "Heatmap.png"),
  height = 8, width = 10,
  plot = Heatmap
)
```

### Principal Component Analysis (PCA)

Dimensionality reduction for evaluating outliers and global sample clusters for quality control and primary exploratory analysis.

```{r}
# PCA plot
pcaRes <- prcomp(t(assay(counts_vst_normalized)))
varExp <- round(summary(pcaRes)[[1]], digits = 2)

pcaDF <- data.frame(
  PC1 = pcaRes$x[, 1],
  PC2 = pcaRes$x[, 2],
  Group = sampleTable$group,
  Sample = sampleTable$sample_ID,
  Dose = sampleTable$dose
)

pcaPlot <- ggplot(pcaDF, mapping = aes(x = PC1, y = PC2, color = Group, label = Sample)) +
  geom_point(aes(shape = Group), size = 3) +
  labs(
    x = paste0("PC1 (", varExp[1], " %)"),
    y = paste0("PC2 (", varExp[2], " %)")
  ) +
  theme(
    axis.text = element_text(size = 12),
    legend.text = element_text(size = 10)
  ) +
  scale_color_manual(values = brewer.pal(3, "Accent")) +
  cowplot::theme_cowplot()

print(pcaPlot)

# Scree plot, showing contribution of principal components in descending order
pca.var <- pcaRes$sdev^2
pca.var.per <- round(pca.var / sum(pca.var) * 100, 1)

barplot(pca.var.per,
  main = "Scree Plot",
  xlab = "Principal Component",
  ylab = "Percent Variation"
)

# Explore which variables contribute the most for each PC
var <- get_pca_var(pcaRes)
var$contrib %>%
  as.data.frame() %>%
  arrange(desc(Dim.1)) %>%
  head() %>%
  kableExtra::kable() %>%
  kable_styling(bootstrap_options = "striped", full_width = FALSE)
```

```{r, echo=FALSE, results='hide'}
# Saving PCA plot
ggsave(
  filename = paste0("../", figures.dir, "PCA.pdf"),
  height = 8, width = 10,
  plot = pcaPlot
)

# Saving Scree plot
{
  pdf(paste0("../", figures.dir, "Scree_plot.pdf"),
    width = 10, height = 8, compress = TRUE
  )
  barplot(pca.var.per,
    main = "Scree Plot",
    xlab = "Principal Component",
    ylab = "Percent Variation"
  )
  par(mfrow = c(1, 1))
  dev.off()
}
```

## Differential Gene Expression Analysis

Differential Gene Expression Analysis using DESeq2. Comparing parameters; dose, time point, prior infection status.

```{r, results='hide'}
# Run the DESeq2 analysis
dds1 <- DESeq(dds1)
dds2 <- DESeq(dds2)
dds3 <- DESeq(dds3)
ddsTC <- DESeq(dds4, test = "LRT", reduced = ~ dose + group)

# Contrast
resultsNames(dds1)
resultsNames(dds2)
resultsNames(dds3)
resultsNames(ddsTC)

# Create function for makings comparisons with default being conditions and dds1
compare_DEGs <- function(variable1, variable2, condition = "conditions", dds = dds1) {
  as.data.frame(results(dds, contrast = c(condition, variable1, variable2)))
}

# Baselines DEGs
base_pos_neg_d1 <- compare_DEGs("pos_0_dose1", "neg_0_dose1")
base_pos_neg_d2 <- compare_DEGs("pos_0_dose2", "neg_0_dose2")
base_pos_neg_d3 <- compare_DEGs("pos_0_dose3", "neg_0_dose3")

base_pos_d1_d2 <- compare_DEGs("pos_0_dose2", "pos_0_dose1")
base_pos_d1_d3 <- compare_DEGs("pos_0_dose3", "pos_0_dose1")
base_pos_d2_d3 <- compare_DEGs("pos_0_dose2", "pos_0_dose3")

base_neg_d1_d2 <- compare_DEGs("neg_0_dose2", "neg_0_dose1")
base_neg_d1_d3 <- compare_DEGs("neg_0_dose3", "neg_0_dose1")
base_neg_d2_d3 <- compare_DEGs("neg_0_dose2", "neg_0_dose3")

# 0-24h and 0-48 (3rd dose) DEGs
neg_d1 <- compare_DEGs("neg_24_dose1", "neg_0_dose1")
neg_d2 <- compare_DEGs("neg_24_dose2", "neg_0_dose2")
neg_d3_24 <- compare_DEGs("neg_24_dose3", "neg_0_dose3")
neg_d3_48 <- compare_DEGs("neg_48_dose3", "neg_0_dose3")

pos_d1 <- compare_DEGs("pos_24_dose1", "pos_0_dose1")
pos_d2 <- compare_DEGs("pos_24_dose2", "pos_0_dose2")
pos_d3_24 <- compare_DEGs("pos_24_dose3", "pos_0_dose3")
pos_d3_48 <- compare_DEGs("pos_48_dose3", "pos_0_dose3")

# Results dds2
pos <- compare_DEGs("pos_24", "pos_0", condition = "time", dds = dds2)
neg <- compare_DEGs("neg_24", "neg_0", condition = "time", dds = dds2)

# Results dds3
d1 <- compare_DEGs("24_dose1", "0_dose1", condition = "dose", dds = dds3)
d2 <- compare_DEGs("24_dose2", "0_dose2", condition = "dose", dds = dds3)
d3_24 <- compare_DEGs("24_dose3", "0_dose3", condition = "dose", dds = dds3)
d3_48 <- compare_DEGs("48_dose3", "0_dose3", condition = "dose", dds = dds3)

# Results dds4 intercept
resTC <- results(ddsTC)
```

### Volcano plots
Volcano plot for data exploratory analysis, filtering for significant values with False Discovery Rate < 0.05 and log fold change greater than 1.

```{r, fig.width=18, fig.height=12, fig.keep=10:17}
results_list <- list(
  base_pos_neg_d1,
  base_pos_neg_d2,
  base_pos_neg_d3,
  base_pos_d1_d2,
  base_pos_d1_d3,
  base_pos_d2_d3,
  base_neg_d1_d2,
  base_neg_d1_d3,
  base_neg_d2_d3,
  neg_d1,
  neg_d2,
  neg_d3_24,
  neg_d3_48,
  pos_d1,
  pos_d2,
  pos_d3_24,
  pos_d3_48,
  pos,
  neg,
  d1,
  d2,
  d3_24,
  d3_48
)

names(results_list) <- c(
  "Baseline Dose 1 Conv vs NaÃ¯ve",
  "Baseline Dose 2 Conv vs NaÃ¯ve",
  "Baseline Dose 3 Conv vs NaÃ¯ve",
  "Baseline Conv Dose 1 and 2",
  "Baseline Conv Dose 1 and 3",
  "Baseline Conv Dose 2 and 3",
  "Baseline NaÃ¯ve Dose 1 and 2",
  "Baseline NaÃ¯ve Dose 1 and 3",
  "Baseline NaÃ¯ve Dose 2 and 3",
  "NaÃ¯ve Dose 1 0-24 hours",
  "NaÃ¯ve Dose 2 0-24 hours",
  "NaÃ¯ve Dose 3 0-24 hours",
  "NaÃ¯ve Dose 3 0-48 hours",
  "Conv Dose 1 0-24 hours",
  "Conv Dose 2 0-24 hours",
  "Conv Dose 3 0-24 hours",
  "Conv Dose 3 0-48 hours",
  "Conv 0-24 hours",
  "NaÃ¯ve 0-24 hours",
  "Dose 1 0-24 hours",
  "Dose 2 0-24 hours",
  "Dose 3 0-24 hours",
  "Dose 3 0-48 hours"
)

results_list_plot <- lapply(results_list, function(x) {
  x <- x %>% mutate(
    test_padj = case_when(
      padj < 0.05 & log2FoldChange >= 1 ~ "Up regulated",
      padj < 0.05 & log2FoldChange <= -1 ~ "Down regulated",
      TRUE ~ "Not significant"
    ),
    test_pvalue = case_when(
      pvalue < 0.05 & log2FoldChange >= 1 ~ "Up regulated",
      pvalue < 0.05 & log2FoldChange <= -1 ~ "Down regulated",
      TRUE ~ "Not significant"
    ),
  )
})

# function to plot volcano plots
volcano_plot <- function(x, pvalue_selection = c("padj", "pvalue")) {
  if (pvalue_selection == "padj") {
    col_name_pval <- "test_padj"
  } else if (pvalue_selection == "pvalue") {
    col_name_pval <- "test_pvalue"
  } else {
    stop("Please change the pvalue_selection argument value. It should either `padj` or `pvalue`.")
  }
  subsetted <- subset(results_list_plot[[x]] %>% tibble::rownames_to_column("gene_symbol"), abs(log2FoldChange) >= 1 & get(pvalue_selection) < 0.05)
  label_list <- c("RSAD2", "IFIT1", "CXCL10", "CMPK2", "GBP1P1", "IFI44L", "BATF2", "SIGLEC1", "IFI44", "NDUFS4", "MX1", "CXCL9", "FCGR1A", "FCGR1B", "CCL4")
  label <- subsetted %>% filter(gene_symbol %in% label_list)

  # count up regulated DEGs
  deg_number_up <- subset(
    results_list_plot[[x]] %>%
      tibble::rownames_to_column("gene_symbol"),
    log2FoldChange >= 1 & get(pvalue_selection) < 0.05
  ) %>%
    nrow()
  # count down regulated DEGs
  deg_number_down <- subset(
    results_list_plot[[x]] %>%
      tibble::rownames_to_column("gene_symbol"),
    log2FoldChange <= -1 & get(pvalue_selection) < 0.05
  ) %>%
    nrow()


  results_list_plot[[x]] %>%
    ggplot(aes(x = log2FoldChange, y = -log10(get(pvalue_selection)))) +
    geom_point(aes(color = get(col_name_pval)), size = 3, alpha = 0.3) +
    scale_color_manual(values = c("Down regulated" = "blue", "Not significant" = "grey80", "Up regulated" = "red")) +
    xlab(expression("Fold Change (Log"[2] * ")")) +
    ylab(expression(paste0("-Log"[10] * "(", pvalue_selection, ")"))) +
    labs(x = NULL, y = NULL) +
    geom_vline(xintercept = c(-1), linetype = "dotted", size = 1) +
    geom_vline(xintercept = c(1), linetype = "dotted", size = 1) +
    geom_hline(yintercept = -log10(0.05), linetype = "dotted", size = 1) +
    geom_label_repel(data = label, aes(log2FoldChange, -log10(get(pvalue_selection)), label = gene_symbol)) +
    xlim(-2, 6) +
    ylim(0, 20) +
    ggtitle(x) +
    annotate(geom = "text", colour = "red", size = 10, x = 4, y = 20, hjust = 0, label = paste0(deg_number_up)) +
    annotate(geom = "text", colour = "blue", size = 10, x = -2, y = 20, hjust = 0, label = paste0(deg_number_down)) +
    ggprism::theme_prism() +
    theme(axis.ticks = element_line(size = .5), legend.position = "none")
}


volcano_plots <- lapply(names(results_list_plot), volcano_plot, pvalue_selection = "padj")
names(volcano_plots) <- names(results_list_plot)
volcano_plots
```


```{r, echo=FALSE, results='hide'}
# save volcano plots
lapply(names(volcano_plots), function(x) {
  ggsave(
    filename = paste0("../", figures.dir, x, "_volcano_plot.pdf"),
    width = 7, height = 7,
    plot = volcano_plots[[x]]
  )
})


# save degs in a csv file

name <- function(x) {
  tibble::rownames_to_column(x, "gene")
}

results_list_named <- lapply(results_list, name)

rbindlist(results_list_named, idcol = "comparison") %>%
  mutate(FC_greater_1 = ifelse(log2FoldChange > 1, 1, 0)) %>%
  relocate(comparison, FC_greater_1) %>%
  filter(padj < 0.05) %>%
  write.csv(., paste0("../", result.dir, "DEGs_all_comparisons.csv"), row.names = TRUE)
```
### Enrichment analysis

Selecting two different databases for search for enrichment analysis. The databases selected were "GO_Biological_Process_2021" and "KEGG_2021_Human".

Also testing the Blood Transcriptome modules presented in [Li et al. 2014](10.1038/ni.2789) using some GSEA package.

```{r, results='hide'}
# Enrichment analysis
listEnrichrDbs()

gmt <- read.gmt("../data/blood_transcription_module/BTM_for_GSEA_20131008.gmt")
# gmt.hallmark <- read.gmt("../data/metabolic_transcriptionmodule/MgSig_Hallmark.gmt")
gmt.reactome <- read.gmt("../data/metabolic_transcriptionmodule/MSigDB_reactome.gmt")
```

### Gene Set Enrichment Analysis

GSEA looks for patterns of enriched genes which are related to each other. Compared with over-representation analysis this does not rely solely on DEGs. Therefore, this method is better when the differences in gene expression is smaller.

#### Annotations

```{r, results='hide'}
# Set desired organim to homo sapiens
organism <- "org.Hs.eg.db"
# BiocManager::install(organism, character.only = TRUE)
library(organism, character.only = TRUE)
```

#### Prepare Input

```{r}
process_to_gsea <- function(x) {
  original_gene_list <- x$log2FoldChange
  names(original_gene_list) <- rownames(x)
  gene_list <- na.omit(original_gene_list)
  gene_list <- sort(gene_list, decreasing = TRUE)
}

ls_processed <- lapply(results_list, process_to_gsea)
```

#### Gene Set Enrichment Li et al BTMs

```{r, include=FALSE}
list_gse <- lapply(ls_processed, function(x) {
  GSEA(
    geneList = x,
    TERM2GENE = gmt,
    minGSSize = 3,
    maxGSSize = 800,
    nPermSimple = 10000,
    pvalueCutoff = 0.05,
    verbose = TRUE,
    pAdjustMethod = "fdr"
  )
})
```

#### Gene Set Enrichment MSigDb Reactome gene sets
```{r, eval==FALSE}
list_gse_reactome <- lapply(ls_processed, function(x) {
  GSEA(
    geneList = x,
    TERM2GENE = gmt.reactome,
    minGSSize = 3,
    maxGSSize = 800,
    nPermSimple = 10000,
    pvalueCutoff = 0.05,
    verbose = TRUE,
    pAdjustMethod = "fdr"
  )
})
```

#### Combined GSEA heatmap

Using the Blood Transcriptome modules presented in [Li et al. 2014](10.1038/ni.2789).

```{r, gsea_heatmap}
my_palette <- colorRampPalette(c("#265891", "white", "#B80F20"))(n = 100)
sel_list_gse <- list_gse[c(
  "NaÃ¯ve Dose 1 0-24 hours",
  "NaÃ¯ve Dose 2 0-24 hours",
  "NaÃ¯ve Dose 3 0-24 hours",
  "NaÃ¯ve Dose 3 0-48 hours",
  "Conv Dose 1 0-24 hours",
  "Conv Dose 2 0-24 hours",
  "Conv Dose 3 0-24 hours",
  "Conv Dose 3 0-48 hours"
)]

combined_gsea <- data.table::rbindlist(lapply(sel_list_gse, as.data.frame), idcol = TRUE) %>%
  group_by(.id) %>%
  arrange(desc(NES)) %>%
  ungroup()

to_heatmap <- combined_gsea %>%
  select(.id, ID, NES) %>%
  filter(NES > 2 | NES < -2, !grepl("TBA", ID)) %>%
  tidyr::pivot_wider(values_from = NES, names_from = ID) %>%
  mutate(across(where(anyNA), ~ tidyr::replace_na(., 0))) %>%
  t()
colnames(to_heatmap) <- to_heatmap[1, ]
to_heatmap <- to_heatmap[-1, ]

col_order <- c(
  "Conv Dose 1 0-24 hours",
  "NaÃ¯ve Dose 1 0-24 hours",
  "Conv Dose 2 0-24 hours",
  "NaÃ¯ve Dose 2 0-24 hours",
  "Conv Dose 3 0-24 hours",
  "NaÃ¯ve Dose 3 0-24 hours",
  "Conv Dose 3 0-48 hours",
  "NaÃ¯ve Dose 3 0-48 hours"
)

to_heatmap <- to_heatmap[, col_order]

BTM_heatmap <- matrix(as.numeric(to_heatmap), ncol = ncol(to_heatmap)) %>%
  pheatmap::pheatmap(
    width = 6,
    height = 20,
    cellwidth = 20,
    cellheight = 5,
    cluster_rows = TRUE,
    cluster_cols = FALSE,
    clustering_method = "complete",
    angle_col = 45,
    border_color = "black",
    cutree_cols = 4,
    color = my_palette,
    labels_col = colnames(to_heatmap),
    labels_row = rownames(to_heatmap),
    treeheight_col = 1,
    fontsize_row = 6,
    fontsize_col = 8,
    gaps_col = c(2, 4, 6)
  )
```

```{r, echo=FALSE, results='hide'}
# save BTM heatmap
ggsave(
  filename = paste0("../", figures.dir, "BTM_heatmap.pdf"),
  height = 8, width = 10,
  plot = BTM_heatmap
)
```


## Fold change comparison between dose 2 and dose 3 

```{r, jaccard_index}
extract_gene_names <- function(x) {
  subsetted <- results_list_plot[[x]] %>%
    tibble::rownames_to_column("gene_symbol") %>%
    select(gene_symbol, log2FoldChange, padj)
  return(subsetted)
}

gene_names <- lapply(names(results_list_plot), extract_gene_names)
names(gene_names) <- names(results_list_plot)

gene_names_filt <- gene_names[c(
  "NaÃ¯ve Dose 1 0-24 hours", "NaÃ¯ve Dose 2 0-24 hours",
  "NaÃ¯ve Dose 3 0-24 hours", "NaÃ¯ve Dose 3 0-48 hours",
  "Conv Dose 1 0-24 hours", "Conv Dose 2 0-24 hours",
  "Conv Dose 3 0-24 hours", "Conv Dose 3 0-48 hours"
)]

gene_names_df_filt <- rbindlist(gene_names_filt, fill = TRUE, idcol = "group_timepoint") %>%
  mutate(
    timepoint = case_when(
      grepl("Dose 1", group_timepoint) ~ "Dose 1",
      grepl("Dose 2", group_timepoint) ~ "Dose 2",
      grepl("Dose 3 0-24", group_timepoint) ~ "Dose 3",
      grepl("Dose 3 0-48", group_timepoint) ~ "Dose 3 48h"
    ),
    group = case_when(
      grepl("Conv", group_timepoint) ~ "Conv",
      grepl("NaÃ¯ve", group_timepoint) ~ "NaÃ¯ve"
    )
  )

gene_names_sign_df <- gene_names_df_filt %>%
  select(-group_timepoint) %>%
  pivot_wider(names_from = group, values_from = c(padj, log2FoldChange), names_glue = "{.value}_{group}") %>%
  mutate(
    # Categorize genes into 'degs_presence' based on significance and regulation
    degs_presence = case_when(
      padj_NaÃ¯ve < 0.05 & padj_Conv < 0.05 & (log2FoldChange_Conv * log2FoldChange_NaÃ¯ve) > 0 ~ "Both significant",
      padj_NaÃ¯ve < 0.05 ~ "NaÃ¯ve only",
      padj_Conv < 0.05 ~ "Conv only",
      TRUE ~ "non_significant"
    ),
    regulation = case_when(
      log2FoldChange_Conv > 0 & log2FoldChange_NaÃ¯ve > 0 ~ "Upregulated",
      log2FoldChange_Conv < 0 & log2FoldChange_NaÃ¯ve < 0 ~ "Downregulated",
      log2FoldChange_Conv > 0 & log2FoldChange_NaÃ¯ve < 0 ~ "Upregulated Conv",
      log2FoldChange_Conv < 0 & log2FoldChange_NaÃ¯ve > 0 ~ "Upregulated NaÃ¯ve"
    )
  )
```

```{r}
filtered_gene_doses <- gene_names_sign_df %>%
  filter(!timepoint %in% c("Dose 1", "Dose 3 48h"))

g1 <- filtered_gene_doses %>%
  filter(degs_presence != "non_significant") %>%
  ggplot(aes(log2FoldChange_NaÃ¯ve, log2FoldChange_Conv, color = degs_presence)) +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "non_significant"), alpha = .3, color = "grey80") +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "NaÃ¯ve only"), alpha = .3) +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "Conv only"), alpha = .3) +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "Both significant"), alpha = .3) +
  scale_color_viridis_d() +
  scale_y_continuous(limits = c(-6, 6)) +
  scale_x_continuous(limits = c(-6, 6)) +
  geom_smooth(data = subset(filtered_gene_doses, degs_presence != "non_significant", se = TRUE), method = "lm") +
  scale_color_npg() +
  stat_cor(method = "pearson", cor.coef.name = "r", p.accuracy = 0.001) +
  facet_wrap(~timepoint) +
  theme_prism(base_line_size = .5, base_fontface = "plain") +
  theme(
    axis.ticks = element_line(size = .5), legend.position = "top",
    aspect.ratio = 1
  )

g1

ggsave(g1,
  filename = paste0("../", result.dir, "linear_relationship_fold_change_main.pdf"),
  height = 7, width = 7
)
```

## Supplemental figure fold change comparison

```{r}
filtered_gene_doses <- gene_names_sign_df %>%
  filter(timepoint %in% c("Dose 1", "Dose 3 48h"))

g2 <- filtered_gene_doses %>%
  filter(degs_presence != "non_significant") %>%
  ggplot(aes(log2FoldChange_NaÃ¯ve, log2FoldChange_Conv, color = degs_presence)) +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "non_significant"), alpha = .3, color = "grey80") +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "NaÃ¯ve only"), alpha = .3) +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "Conv only"), alpha = .3) +
  geom_point(data = subset(filtered_gene_doses, degs_presence == "Both significant"), alpha = .3) +
  scale_color_viridis_d() +
  scale_y_continuous(limits = c(-6, 6)) +
  scale_x_continuous(limits = c(-6, 6)) +
  geom_smooth(data = subset(filtered_gene_doses, degs_presence != "non_significant", se = TRUE), method = "lm") +
  scale_color_npg() +
  stat_cor(method = "pearson", cor.coef.name = "r", p.accuracy = 0.001) +
  facet_wrap(~timepoint) +
  theme_prism(base_line_size = .5, base_fontface = "plain") +
  theme(
    axis.ticks = element_line(size = .5), legend.position = "top",
    aspect.ratio = 1
  )

g2

ggsave(g2,
  filename = paste0("../", result.dir, "linear_relationship_fold_change_supp.pdf"),
  height = 7, width = 7
)
```

```{r}
degs <- lapply(gene_names_filt, function(x) {
  x %>%
    filter(padj < 0.05, log2FoldChange > 1) %>%
    pull(gene_symbol)
})
degs <- degs[-c(1, 4, 8)]

gom_obj <- newGOM(degs, degs)

filtered_degs_matrix <- getMatrix(gom_obj, "Jaccard")
```

## Jaccard index of differentially expressed genes

```{r}
my_palette <- colorRampPalette(c("#265891", "#F6FF92", "#B80F20"))(n = 50)
# order matrix based on cluster
od <- hclust(dist(filtered_degs_matrix), method = "ward.D2")$order
filtered_degs_matrix_ordered <- filtered_degs_matrix[od, od]

# Define color values for different row annotations
row_col_group <- c("NaÃ¯ve" = "#A9A9A9", "Conv" = "#A9CAF6")
row_col_time <- c(
  "Dose 1 0-24 hours" = "#5e4fa2",
  "Dose 2 0-24 hours" = "#fdae61",
  "Dose 3 0-24 hours" = "#9e0142",
  "Dose 3 0-48 hours" = "#abdda4"
)

# Create row annotations for heatmap
row_anno <- rowAnnotation(
  timepoint = gsub(".*? (.*)$", "\\1", rownames(filtered_degs_matrix_ordered)),
  group = gsub("^(.*?) .*", "\\1", rownames(filtered_degs_matrix_ordered)),
  col = list(timepoint = row_col_time, group = row_col_group),
  annotation_legend_param = list(
    timepoint = list(title = "Timepoint"),
    group = list(title = "Group")
  )
)

col_anno <- columnAnnotation(
  timepoint = gsub(".*? (.*)$", "\\1", rownames(filtered_degs_matrix_ordered)),
  group = gsub("^(.*?) .*", "\\1", rownames(filtered_degs_matrix_ordered)),
  col = list(timepoint = row_col_time, group = row_col_group),
  show_legend = FALSE
)

heatmap1 <- Heatmap(filtered_degs_matrix_ordered,
  border = FALSE,
  rect_gp = gpar(type = "none"),
  show_row_names = FALSE, show_column_names = FALSE,
  name = "Jaccard similarity index",
  bottom_annotation = col_anno,
  right_annotation = row_anno,
  col = my_palette,
  clustering_method_row = "ward.D2",
  clustering_method_columns = "ward.D2",
  show_column_dend = FALSE,
  cell_fun = function(j, i, x, y, w, h, fill) {
    if (as.numeric(x) <= 1.1 - as.numeric(y)) {
      grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
    }
  }
)
heatmap1
{
  pdf(file = paste0("../", result.dir, "jaccard_index_plot.pdf"), height = 5, width = 7)
  draw(heatmap1)
  dev.off()
}
```

## Cytokine heatmap

4 gene sets were used based on the groups of  Chemokine (45), Interferon (33), Interleukins (43), and TNF (18) ligands according to the HUGO gene nomenclature Committee.

```{r}
ls_gene_file <- list.files("../data/gene_sets", pattern = "231102|231128", full.names = TRUE)
ls_gene_sets <- lapply(ls_gene_file, read.csv, header = TRUE, sep = ",")
names(ls_gene_sets) <- sub("^(?:[^_]+_){2}([^_]+)_.*$", "\\1", ls_gene_file)
gene_sets <- rbindlist(ls_gene_sets, idcol = "gene_set", use.names = TRUE)

cytokine_heatmap <- gene_names_sign_df %>%
  pivot_longer(
    cols = starts_with("padj_") | starts_with("log2FoldChange_"),
    names_to = c(".value", "group"),
    names_pattern = "([\\p{L}]+)_(\\w+)"
  ) %>%
  filter(gene_symbol %in% gene_sets$Approved_symbol) %>%
  mutate(
    group = ifelse(group == "Na", "NaÃ¯ve", group),
    group_timepoint = paste(timepoint, group)
  ) %>%
  group_by(gene_symbol, timepoint) %>%
  mutate(FoldChange_mean = case_when(
    degs_presence == "Both significant" ~ mean(FoldChange),
    degs_presence == "non_significant" ~ min(FoldChange),
    degs_presence %in% c("Conv only", "NaÃ¯ve only") ~ max(FoldChange)
  )) %>%
  ungroup()

g2 <- cytokine_heatmap %>%
  group_by(gene_symbol) %>%
  filter(
    !all(degs_presence == "non_significant"),
    regulation %in% c("Upregulated", "Downregulated")
  ) %>%
  ungroup() %>%
  ggplot(aes(x = timepoint, y = gene_symbol, size = abs(FoldChange_mean), color = degs_presence)) +
  geom_point() +
  scale_color_manual(values = c("#D95F02", "#A9CAF6", "#A9A9A9", "grey90")) +
  labs(x = "", y = "", size = "FoldChange\n(Absolute log2)", color = "Significance") +
  theme_bw() +
  facet_wrap(~regulation, scales = "free_y") +
  theme_prism(base_line_size = .5, base_fontface = "plain", axis_text_angle = 45, base_size = 12) +
  theme(
    axis.ticks = element_line(size = .5), legend.position = "right",
    aspect.ratio = 1,
    legend.title = element_text()
  )
g2
ggsave(g2,
  filename = paste0("../", result.dir, "cytokine_degs.pdf"),
  height = 7, width = 7
)
```

### Generating examples of cytokine DEGs

```{r degs_show}
# interesting genes related to complement
plot_counts <- function(gene_interest, dds = dds1) {
  fiss <- plotCounts(dds, gene_interest,
    intgroup = c("dose", "group"), returnData = TRUE, transform = FALSE, normalized = TRUE, replaced = FALSE
  )

  g <- ggplot(
    fiss,
    aes(x = dose, y = count, color = group, group = group)
  ) +
    geom_point() +
    stat_summary(fun.y = mean, geom = "line") +
    scale_y_log10() +
    labs(x = "", y = "Normalized counts\n(median of ratios)") +
    scale_color_manual(values = c("#A9A9A9", "#A9CAF6")) +
    theme_prism(base_line_size = .5, base_fontface = "plain", axis_text_angle = 45) +
    theme(
      axis.ticks = element_line(size = .5), legend.position = "right",
      aspect.ratio = 1,
      legend.title = element_text()
    )
  g

  ggsave(g,
    filename = paste0("../", result.dir, gene_interest, "_cytokine_degs.pdf"),
    height = 5, width = 5
  )
}
plot_counts(gene_interest = "CXCL10")

plot_counts(gene_interest = "IL16")

plot_counts(gene_interest = "CD40LG")

plot_counts(gene_interest = "CCL5")

plot_counts(gene_interest = "TNF")

plot_counts(gene_interest = "TNFSF13B")

plot_counts(gene_interest = "IL1B")

plot_counts(gene_interest = "IL15")
```

## Transcription factors heatmap
overrepresentation analysis of transcriptional factors

```{r}
msig_geneset <- msigdbr(species = "Homo sapiens", category = "C3")
msig_geneset <- msig_geneset %>% filter(gs_subcat %in% c("TFT:GTRD", "TFT:TFT_Legacy"))
tft_geneset <- msig_geneset %>%
  dplyr::select(gs_name, gene_symbol) %>%
  dplyr::rename(term = gs_name, gene = gene_symbol)

annot <- ensembldb::select(EnsDb.Hsapiens.v86,
  keys = keys(EnsDb.Hsapiens.v86),
  columns = c("ENTREZID", "SYMBOL")
)

degs <- lapply(degs, function(x) plyr::mapvalues(x, annot$GENEID, annot$SYMBOL, warn_missing = FALSE))

## map names with TFs
ora_tf <- lapply(degs, function(x) {
  enricher(x,
    pvalueCutoff = 1,
    pAdjustMethod = "fdr",
    qvalueCutoff = 1,
    TERM2GENE = tft_geneset
  )
})

ora_tf_df <- lapply(ora_tf, data.frame)
ora_tf_df_bind <- rbindlist(ora_tf_df, use.names = TRUE, idcol = "group_timepoint", fill = TRUE) %>% as.data.frame()
```

### Plot TF data
```{r}
ora_tf_df_bind <- ora_tf_df_bind %>%
  mutate(
    timepoint = case_when(
      grepl("Dose 1", group_timepoint) ~ "Dose 1",
      grepl("Dose 2", group_timepoint) ~ "Dose 2",
      grepl("Dose 3 0-24", group_timepoint) ~ "Dose 3",
      grepl("Dose 3 0-48", group_timepoint) ~ "Dose 3 48h"
    ),
    group = case_when(
      grepl("Conv", group_timepoint) ~ "Conv",
      grepl("NaÃ¯ve", group_timepoint) ~ "NaÃ¯ve"
    ),
    timepoint_group = paste(timepoint, group, sep = "_")
  )

# Create the ggplot with the reordered x-axis

ora_tf_df_bind <- ora_tf_df_bind %>%
  filter(p.adjust < 0.05 & Count > 1) %>%
  mutate(
    ID_cut = case_when(
      grepl("TARGET", ID) ~ sub("_.*$", "", ID),
      str_count(ID, "_") > 1 ~ sub("^[^_]*_(.*?)_.*$", "\\1", ID),
      TRUE ~ sub("_.*$", "", ID),
    ),
    ID_cut = ifelse(ID_cut == "ICSBP", "IRF8/ICSBP", ID_cut)
  ) %>%
  filter(!(ID_cut %in% c("YNGTTNNNATT", "NFKAPPAB65", "NFKB"))) %>%
  # Remove duplicated ID by selecting the gene set with lowest p_value
  arrange(ID, p.adjust) %>%
  distinct(ID, timepoint_group, .keep_all = TRUE)

# create empty data.frame with Naive Dose 1 for aesthetics
data_empty <- data.frame(
  timepoint_group = "Dose 1_NaÃ¯ve",
  ID_cut = unique(ora_tf_df_bind$ID_cut),
  p.adjust = 1
)

ora_tf_df_bind <- plyr::rbind.fill(ora_tf_df_bind, data_empty)

g3 <- ora_tf_df_bind %>%
  ggplot(aes(x = timepoint_group, y = ID_cut, fill = -log10(p.adjust))) +
  geom_tile() +
  theme_bw() +
  scale_fill_gradient2(low = "#265891", mid = "white", high = "#B80F20") +
  labs(x = "", y = "", fill = "Adjusted p-value\n(-log10)") +
  theme_prism(base_line_size = .5, base_fontface = "plain", axis_text_angle = 45) +
  theme(
    axis.ticks = element_line(size = .5), legend.position = "right",
    aspect.ratio = 1,
    legend.title = element_text()
  )
g3
ggsave(g3,
  filename = paste0("../", result.dir, "transcription_factors_ORA.pdf"),
  height = 7, width = 7
)
```

## ORA tf cluster comparisons

```{r}
degs[["NaÃ¯ve Dose 1 0-24 hours"]] <- "no_degs"
degs_entrez_id <- lapply(degs, function(x) plyr::mapvalues(x, from = annot$SYMBOL, to = annot$ENTREZID, warn_missing = FALSE))

compared_cluster <- compareCluster(degs_entrez_id, fun = "enrichKEGG", organism = "hsa")
cluster_termsim <- pairwise_termsim(compared_cluster, method = "JC")

# create empty data.frame with Naive Dose 1 for aesthetics
data_empty <- data.frame(
  Cluster = "NaÃ¯ve Dose 1 0-24 hours",
  ID = cluster_termsim@compareClusterResult$ID,
  Description = cluster_termsim@compareClusterResult$Description,
  log10_padj = 0,
  pvalue = 1,
  p.adjust = 1,
  qvalue = 1,
  geneID = 0,
  Count = 1, GeneRatio = 0, BgRatio = 0
) %>% distinct()
cluster_termsim@compareClusterResult <- cluster_termsim@compareClusterResult %>%
  mutate(log10_padj = -log10(p.adjust))

cluster_termsim@compareClusterResult <- plyr::rbind.fill(cluster_termsim@compareClusterResult, data_empty)

# Set factor levels
cluster_termsim@compareClusterResult$Cluster <- factor(cluster_termsim@compareClusterResult$Cluster, levels = c(
  "Conv Dose 1 0-24 hours", "NaÃ¯ve Dose 1 0-24 hours",
  "Conv Dose 2 0-24 hours", "NaÃ¯ve Dose 2 0-24 hours",
  "Conv Dose 3 0-24 hours", "NaÃ¯ve Dose 3 0-24 hours"
))


treeplot(cluster_termsim,
  showCategory = 31,
  label_format = 30,
  offset.params = list(
    tiplab = 9,
    bar_tree = 23
  ),
  cluster.params = list(
    n = 6,
    label_words_n = 0
  ),
  clusterPanel.params = list(
    colnames_angle = -90
  ),
  color = "log10_padj"
) +
  labs(x = "", fill = "Adjusted p-value\n(-log10)") +
  scale_fill_gradient2(low = "#265891", mid = "white", high = "#B80F20") +
  theme(
    axis.ticks = element_line(size = .5), legend.position = "right",
    aspect.ratio = 1,
    legend.title = element_text()
  )

ggsave(
  filename = paste0("../", result.dir, "ora_kegg_pathway_tree.pdf"),
  height = 10, width = 30
)
```

## Vaccine detection in RNA-seq

```{r}
tpm_gene <- data.table::fread("../data/processed_data_nfcore/batch_1-2/salmon/salmon.merged.gene_tpm.tsv", header = TRUE, sep = "\t") %>%
  as.data.frame()

colnames(tpm_gene) <- plyr::mapvalues(sub("_L001|_L002", "", colnames(tpm_gene)), from = sampleTable$sample_ID, to = paste(sampleTable$subject_ID, sampleTable$dose, sampleTable$group, sep = "_"))

names_col <- colnames(tpm_gene)[3:101]
vaccine_tpm <- tpm_gene[64622:64625, ]

vaccine_plot <- tidyr::pivot_longer(vaccine_tpm, cols = names_col) %>%
  tidyr::separate(name, into = c("subject_ID", "time", "dose", "group"), sep = "_", remove = TRUE) %>%
  mutate(
    time_dose = paste(time, dose, sep = "_"),
    time_dose = factor(time_dose, levels = c(
      "0_dose1", "24_dose1",
      "0_dose2", "24_dose2",
      "0_dose3", "24_dose3", "48_dose3"
    )),
    group = case_when(
      group == "pos" ~ "Conv",
      group == "neg" ~ "NaÃ¯ve"
    ),
    group = factor(group, levels = c("Conv", "NaÃ¯ve"))
  ) %>%
  na.omit()

vaccine_plot %>%
  filter(grepl("pfizer", gene_id)) %>%
  ggplot(aes(x = time_dose, y = value, group = subject_ID, fill = group)) +
  geom_line() +
  geom_point(shape = 21, size = 5, color = "black", stroke = 1.2) +
  scale_fill_manual(values = c("#A9CAF6", "#A9A9A9")) +
  scale_y_log10(limits = c(0.001, 1000), breaks = c(0.01, 0.1, 1, 10, 100)) +
  labs(y = "Transcripts per million (log)", x = "") +
  ggprism::theme_prism() +
  theme(
    legend.position = "none",
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  ) +
  facet_wrap(~group, ncol = 4)

ggsave(
  filename = paste0("../", figures.dir, "pfizer_detection_rna-seq.pdf"),
  height = 5, width = 7
)
```



